sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/DelegateAnnotationVisitor.java

```diff
@@ -0,0 +1,251 @@
+
+    @Override
+    public Set<String> getSupportedAnnotationNames() {
+        return Set.of(Delegate.class.getName());
+    }
+
+    @Override
+    public void visitClass(ClassElement element, VisitorContext context) {
+        AnnotationValue<?> annotation = element.getAnnotation(Delegate.class);
+        Optional<Class<?>> type = annotation.classValue("type");
+
+        if (type.isPresent() && !type.get().equals(Void.class)) {
+            ClassElement typeElement = context.getClassElement(type.get()).orElseThrow(
+                () -> new ProcessingException(element, "Could not find required type " + type.get().getName()));
+            createDelegate(typeElement, context);
+        } else {
+            createDelegate(element, context);
+        }
+    }
+
+    private void createDelegate(ClassElement element, VisitorContext context) {
+        if (processed.contains(element.getName())) {
+            return;
+        }
+        if (!element.isInterface()) {
+            throw new ProcessingException(element, "Only interfaces are supported for delegate creation. But '"
+                + element.getName() + "' is annotated with @Delegate");
+        }
+
+        try {
+            String simpleName = element.getSimpleName() + NAME_SUFFIX;
+            String delegateClassName = element.getPackageName() + "." + simpleName;
+
+            ClassTypeDef typeDef = ClassTypeDef.of(element);
+            ClassTypeDef delegateType = ClassTypeDef.of(delegateClassName);
+            ClassDefBuilder delegate = ClassDef.builder(delegateClassName)
+                .addModifiers(Modifier.PUBLIC, Modifier.ABSTRACT);
+
+            if (!element.getTypeArguments().isEmpty()) {
+                for (Entry<String, ClassElement> type: element.getTypeArguments().entrySet()) {
+                    if (type.getValue() instanceof GenericPlaceholderElement placeholderElement) {
+                        delegate.addTypeVariable(new TypeVariable(
+                            type.getKey(),
+                            placeholderElement.getBounds().stream().map(TypeDef::of).toList()
+                        ));
+                    } else {
+                        delegate.addTypeVariable(new TypeVariable(type.getKey()));
+                    }
+                }
+                typeDef = TypeDef.parameterized(
+                    typeDef,
+                    element.getTypeArguments().entrySet().stream().map(
+                        v -> (TypeDef) new TypeVariable(v.getKey())
+                    ).toList()
+                );
+            }
+            delegate.addSuperinterface(typeDef)
+                .addField(FieldDef.builder(DELEGATEE_MEMBER).ofType(typeDef).build())
+                .addAllFieldsConstructor();
+
+            Set<MethodSignature> createdMethods = new HashSet<>();
+            addDelegateMethods(element, delegate, typeDef, delegateType, createdMethods);
+            for (ClassElement interfaceElement : getAllInterfaces(element)) {
```

Comment:
What is the point of getting interfaces? `getMethods` should return all methods from the class